File: BitcoinSwift/SecureData.m
//
//  SecureData.m
//  BitcoinSwift
//
//  Created by Kevin Greene on 1/11/15.
//  Copyright (c) 2015 DoubleSha. All rights reserved.
//

#import "SecureData.h"

#import "SecureMemoryAllocator.h"

@interface SecureData()

@property(nonatomic, strong) NSMutableData *mutableData;

@end

@implementation SecureData

- (instancetype)init {
  self = [super init];
  if (self) {
    _mutableData = CFBridgingRelease(CFDataCreateMutable([SecureMemoryAllocator allocator], 0));
  }
  return self;
}

- (instancetype)initWithLength:(NSUInteger)length {
  self = [super init];
  if (self) {
    _mutableData = CFBridgingRelease(CFDataCreateMutable([SecureMemoryAllocator allocator],
                                                         length));
    _mutableData.length = length;
  }
  return self;
}

- (instancetype)initWithData:(NSData *)data {
  self = [self initWithLength:data.length];
  if (self) {
    memcpy(_mutableData.mutableBytes, data.bytes, data.length);
  }
  return self;
}

- (instancetype)initWithBytes:(const void *)bytes length:(NSUInteger)length {
  self = [self initWithLength:length];
  if (self) {
    memcpy(_mutableData.mutableBytes, bytes, length);
  }
  return self;
}

- (NSData *)data {
  return _mutableData;
}

- (NSMutableData *)mutableData {
  return _mutableData;
}

- (const void *)bytes {
  return _mutableData.bytes;
}

- (void *)mutableBytes {
  return _mutableData.mutableBytes;
}

- (NSUInteger)length {
  return _mutableData.length;
}

- (void)appendBytes:(const void *)bytes length:(NSUInteger)length {
  [_mutableData appendBytes:bytes length:length];
}

- (void)appendData:(NSData *)data {
  [_mutableData appendData:data];
}

- (void)appendSecureData:(SecureData *)secureData {
  [_mutableData appendData:secureData.mutableData];
}

@end

File: BitcoinSwift/SecureData.swift
//
//  SecureData.swift
//  BitcoinSwift
//
//  Created by Kevin Greene on 12/30/14.
//  Copyright (c) 2014 DoubleSha. All rights reserved.
//

import Foundation

public func ==(left: SecureData, right: SecureData) -> Bool {
  return left.mutableData == right.mutableData
}

public class SecureData: Equatable {

  private let mutableData: NSMutableData

  public var mutableBytes: UnsafeMutablePointer<Void> {
    return mutableData.mutableBytes
  }

  public var bytes: UnsafePointer<Void> {
    return mutableData.bytes
  }

  public var length: Int {
    return mutableData.length
  }

  public subscript(subRange: Range<Int>) -> SecureData {
    precondition(mutableData.length > 0)
    precondition(subRange.startIndex >= 0 && subRange.startIndex < mutableData.length)
    precondition(subRange.endIndex > subRange.startIndex &&
        subRange.endIndex <= mutableData.length)
    let length = subRange.endIndex - subRange.startIndex
    let range = NSRange(location: subRange.startIndex, length: length)
    var subData = SecureData(length: length)
    memcpy(subData.mutableBytes, bytes.advancedBy(subRange.startIndex), UInt(length))
    return subData
  }

  public init() {
    mutableData = CFDataCreateMutable(SecureMemoryAllocator.allocator().takeUnretainedValue() , 0)
  }

  public init(length: Int) {
    mutableData =
        CFDataCreateMutable(SecureMemoryAllocator.allocator().takeUnretainedValue(), length)
    mutableData.length = length
  }

  public convenience init(bytes: UnsafePointer<Void>, length: Int) {
    self.init(length: length)
    memcpy(mutableBytes, bytes, UInt(length))
  }
}
